README.md
<p align="center">
  <!-- logo -->
  <img src="https://github.com/edoardottt/images/blob/main/pwdsafety/logo.jpg"><br>
  <b>Command line tool that checks how much a password is safe</b><br>
  <sub>
    Coded with 💙 by edoardottt.
  </sub>
</p>

<!-- badges -->
<p align="center">
    <!-- mainteinance -->
      <a href="https://edoardoottavianelli.it">
        <img src="https://img.shields.io/badge/Maintained%3F-yes-green.svg" alt="Mainteinance yes" />
      </a>
    <!-- version -->
      <a href="https://edoardoottavianelli.it">
        <img src="https://github.com/edoardottt/images/blob/main/pwdsafety/version.svg" alt="version" />
      </a>
    <!-- pr-welcome -->
      <a href="https://edoardoottavianelli.it">
        <img src="https://github.com/edoardottt/READMENATOR/blob/master/images/pr-welcome.svg" alt="pr-welcome" />
      </a>
    <!-- ask-me-anything -->
      <a href="https://edoardoottavianelli.it">
        <img src="https://github.com/edoardottt/READMENATOR/blob/master/images/ask-me-anything.svg" alt="ask me anything" />
      </a>
  <br>
    <!-- go-report-card -->
      <a href="https://goreportcard.com/report/github.com/edoardottt/pwdsafety">
        <img src="https://goreportcard.com/badge/github.com/edoardottt/pwdsafety" alt="go-report-card" />
      </a>
    <!-- workflows -->
      <a href="https://edoardoottavianelli.it">
        <img src="https://github.com/edoardottt/pwdsafety/workflows/Go/badge.svg?branch=master" alt="workflows" />
      </a>
    <!-- ubuntu-build -->
      <a href="https://edoardoottavianelli.it">
        <img src="https://github.com/edoardottt/images/blob/main/pwdsafety/ubuntu-build.svg" alt="ubuntu-build" />
      </a>
  <br>
    <!-- gobadge -->
      <a href="https://edoardoottavianelli.it">
        <img src="https://github.com/edoardottt/images/blob/main/pwdsafety/gobadge" alt="gobadge" />
      </a>
    <!-- license GPLv3.0 -->
      <a href="https://github.com/edoardottt/READMENATOR/blob/master/LICENSE">
        <img src="https://github.com/edoardottt/READMENATOR/blob/master/images/license-GPL3.svg" alt="license-GPL3" />
      </a>
</p>
<p align="center">
  <a href="#example-bar_chart">Example</a> •
  <a href="#get-started-">Get Started</a> •
  <a href="#description-">Description</a> •
  <a href="#scoring-">Scoring</a> •
  <a href="#contributing-">Contributing</a>
</p>

**This tool doesn't store any information!!**  
**Remember, never use personal(or related to you) info in your password!**  
 - Use a password manager (I recommend [BitWarden](https://bitwarden.com/))  
 - Don't use the same password for different services  
 - Enable 2FA when possible  

Example :bar_chart:
----------

[![asciicast](https://asciinema.org/a/406710.svg)](https://asciinema.org/a/406710)

Get Started 🎉
----------

- First of all, clone the repo locally

  - `git clone https://github.com/edoardottt/pwdsafety.git`

- pwdsafety has external dependencies, so they need to be pulled in:

  - `cd pwdsafety`
  - `go get`

- Linux (Requires high perms, run with sudo)

  - `make linux` (to install)

  - `make unlinux` (to uninstall)

- Windows (executable works only in pwdsafety folder. Alias?)

  - `make windows` (to install)

  - `make unwindows` (to uninstall)

Description 🔦 
----------

It reads from standard input the entered password.  
First, it searches if the password or the password reversed is a well known pwd.  
Then, just do little calculations, checking if the basic rules are respected, like if there are UPPERCASE CHARS, lowercase chars, numb3rs and symbols.  
It stores the length of the password and the ratio [ unique different chars / total chars].  
It calculates then the entropy of a password.  
Password entropy is a measurement of how unpredictable a password is.  
The formula for entropy is:  
              ![formula](https://github.com/edoardottt/images/blob/main/pwdsafety/formula.png)  
              
Where:
- E = password entropy  
- R = pool of unique characters  
- L = number of characters in your password  
- Then R^L = the number of possible passwords  

When the score <= 68(reasonable) it generates a random password.  

Scoring 💯
----------

**Max score: 100**

**Scores:**
  - Very weak: 0 - 35
  - Weak: 36 - 59
  - Reasonable: 60 - 68
  - Strong: 69 - 80
  - Very strong: 81 -100
  
**Scoring parameters:**
  - Found in known password
  - Found in known password reversed
  - Password composition:
      - numbers
      - symbols
      - uppercase
      - lowercase
  - Unique different characters
  - Length
  - Entropy

Contributing 🛠
-------

Just open an [issue](https://github.com/edoardottt/pwdsafety/issues) / [pull request](https://github.com/edoardottt/pwdsafety/pulls). 

See also [CONTRIBUTING.md](https://github.com/edoardottt/pwdsafety/blob/master/CONTRIBUTING.md) and [CODE OF CONDUCT.md](https://github.com/edoardottt/pwdsafety/blob/master/CODE_OF_CONDUCT.md)

Thanks to [fabaff](https://github.com/fabaff) and [ecnepsnai](https://github.com/ecnepsnai/pwnedpassword/blob/master/pwned.go).

License 📝
-------

This repository is under [GNU General Public License v3.0](https://github.com/edoardottt/pwdsafety/blob/master/LICENSE).  
[edoardoottavianelli.it](https://www.edoardoottavianelli.it) to contact me.


CHANGELOG.md
**[v0.2](https://github.com/edoardottt/pwd-safety/releases/tag/v0.2):**
          
          - Remove all the txt files containing known password.
          - Now it checks if the password (and the pwd reversed) is leaked via an API.


**[v0.1.3](https://github.com/edoardottt/pwd-safety/releases/tag/v0.1.3):**
          
          - Added:
                  - Hash hex results for password
                  - Better output readability

**[v0.1.2](https://github.com/edoardottt/pwd-safety/releases/tag/v0.1.2):**
          
          - Solved Bad Input from Keyboard
          - Added:
                  - Password Cracking time estimate
                  - Makefile

**[v0.1.1](https://github.com/edoardottt/pwd-safety/releases/tag/v0.1.1):**
  
          - Added:
                  - Generates random password when score <= REASONABLE
                  - Errors code table

**[v0.1](https://github.com/edoardottt/pwd-safety/releases/tag/v0.1):**
  
          - First pre-release


Scoring.go
/*
 *    This program is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU Public License as published
 *    by the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *    Author:
 *      Edoardo Ottavianelli <edoardott@gmail.com>
 */

package main

import (
	"fmt"
	"os"
)

/*
LengthScore :
Scores password's length
	total = 30
	length<=7 = 0
	length==8 = 4
	length==9 = 10
	10 <= length <= 15 = 15
	16 <= length <= 19 = 21
	20 <= length <= 24 = 26
	length >=25 = 30
*/
func LengthScore(password string) float64 {
	length := len(password)
	if length <= 7 {
		return 0
	}
	if length == 8 {
		return 4
	}
	if length == 9 {
		return 10
	}
	if length > 9 && length < 16 {
		return 15
	}
	if length > 15 && length < 20 {
		return 21
	}
	if length > 19 && length < 25 {
		return 26
	}
	return 30
}

/*
CompositionPwdScore :
Scores password's composition
	total = 20
	There is numbers = 5
	There is symbol = 5
	There is uppercase = 5
	There is lowercase = 5
*/
func CompositionPwdScore(password string) float64 {
	var result int
	numbers := IsThereNumber(password)
	upper := IsThereUpperCase(password)
	lower := IsThereLowerCase(password)
	symbols := IsThereSymbol(password)
	if numbers {
		result += 5
	}
	if upper {
		result += 5
	}
	if lower {
		result += 5
	}
	if symbols {
		result += 5
	}
	return float64(result)
}

/*
DifferentCharScore :
Scores How many different chars in relation to the length
	total = 15
	n = (different_chars*total)/total_chars
*/
func DifferentCharScore(password string) float64 {
	diffChars := HowManyDifferents(password)
	total := len(password)
	if total == 0 {
		return 0
	}
	return float64((diffChars * 15) / total)
}

/*
EntropyScore :
Scores Entropy's password
	total = 35
	< 28 bits = 3
	28 - 35 bits = 8
	36 - 59 bits = 20
	60 - 80 bits = 24
	81 - 120 bits = 28
	120+ bits = 35
*/
func EntropyScore(password string) float64 {
	entropy := Entropy(password)
	entropyScore := (entropy * 35) / 130
	if entropyScore > 35 {
		return 35
	}
	return entropyScore
}

/*
pwnedPwds returns the scores for pwned password and
reversed password.

Found : -30
Not Found : 0

Found (Reversed): -10
Not Found (Reversed): 0
*/
func pwnedPwds(password string) (float64, float64) {
	var scoreKnownPwd float64
	var scoreKnownPwdReverse float64
	var knownPwd *Result
	var knownPwdReverse *Result
	//check if password is into known leaked passwords

	knownPwd, err := IsPwned(password)
	if err != nil {
		fmt.Println("Error while retrieving data on password...")
		os.Exit(1)
	}

	knownPwdReverse, err = IsPwned(Reverse(password))
	if err != nil {
		fmt.Println("Error while retrieving data on password...")
		os.Exit(1)
	}

	if knownPwd.Pwned {
		scoreKnownPwd = -30
	} else {
		scoreKnownPwd = 0
	}

	if knownPwdReverse.Pwned {
		scoreKnownPwdReverse = -10
	} else {
		scoreKnownPwdReverse = 0
	}
	return scoreKnownPwd, scoreKnownPwdReverse
}

//Grader : Return the score of the password
func Grader(words [][]string, password string) float64 {
	var optimalLength = 27
	var optimalDifferentCharScore float64 = 7
	var knownStr string
	var knownStrReverse string
	var scoreKnownPwd float64
	var scoreKnownPwdReverse float64

	scoreKnownPwd, scoreKnownPwdReverse = pwnedPwds(password)
	lengthScore := LengthScore(password)
	compositionPwdScore := CompositionPwdScore(password)
	differentCharScore := DifferentCharScore(password)
	entropyScore := EntropyScore(password)
	//Printing results
	if scoreKnownPwd != 0 {
		knownStr = "Yes (-30)"
	} else {
		knownStr = "No"
	}
	if scoreKnownPwdReverse != 0 {
		knownStrReverse = "Yes (-10)"
	} else {
		knownStrReverse = "No"
	}
	fmt.Println("[%] Password found in known leaked passwords: " + knownStr)
	fmt.Println("[%] Password (reversed) found in known leaked passwords: " + knownStrReverse)
	fmt.Println("[%] Length Score: " + fmt.Sprint(lengthScore) + "/30")
	fmt.Println("[%] Composition Score: " + fmt.Sprint(compositionPwdScore) + "/20")
	fmt.Println("[%] Unique chars Score: " + fmt.Sprint(differentCharScore) + "/15")

	entropyRounded := Round(fmt.Sprintf("%.2f", entropyScore))
	fmt.Println("[%] Entropy Score: " + fmt.Sprint(entropyRounded) + "/35")
	score := scoreKnownPwd + scoreKnownPwdReverse + lengthScore + compositionPwdScore + differentCharScore + entropyScore
	//if it's an optimal password by very high length and good different/unique ratio score
	if differentCharScore >= optimalDifferentCharScore && len(password) > optimalLength {
		return 100
	}
	if score > 0 {
		return score
	}
	return 0
}
